index.rst

```diff
@@ -1,61 +1,486 @@
+Having one init quantum per task hints at a solution to another problem: if we write a different software version dataset for each (per-task) init quantum, instead of one for the entire RUN, each can be handled as a regular, write-once dataset, instead of needing to simulate update-in-place behavior.
+
+No predicted-only datasets
+""""""""""""""""""""""""""
+
+Instead of adding the ``dataset.actually_produced`` column to the ``dataset`` table, we could ignore predicted-only datasets entirely in provenance.
+This simplifies the schema and avoids the problem how to query only for datasets that were actually produced.
+These predictions will still appear in ``QuantumGraph`` objects prior to execution, so by dropping them we make it impossible to reconstruct from provenance the full graph that was e.g. submitted to a batch system - but reproducing ``QuantumGraph`` generation is already not a guarantee of the system, because it depends on inputs (e.g. collections) that change over time.
+We would retain the ability to exactly re-execute any particular quantum and reproduce all output datasets however, because by definition these datasets are those that cannot have affected the results of the processing that did occur.
+
+Reverse-lookups for special datasets
+""""""""""""""""""""""""""""""""""""
+
+The provenance tables in the original proposal do not contain any foreign-key columns for the datasets that store important provenance, such as software versions, task configuration, quantum metadata, or quantum logs.
+Instead, these connections rely on dataset type conventions and RUN collection membership, e.g. "the configuration for the task ``<label>`` is in the ``<label>_config`` dataset in the same RUN collection," essentially the same as it is now.
+This avoids circular dependencies between foreign keys (which require an ``UPDATE`` after each ``INSERT`` to fully populate), and it allows the set of provenance-relevant datasets to evolve without schema changes (provided some in-code scheme exists for remembering the historical conventions for provenance dataset types).
+
+It also isn't straightforward to just add dataset foreign key columns to these tables in some cases: the ``task`` table as defined here has rows that are not tied to any particular RUN collection, and hence can't hold a reference to the config dataset.
+
+.. _recording-provenance:
+
+Recording provenance during execution
+-------------------------------------
+
+Avoiding per-dataset or per-quantum communication with a central SQL database is absolutely critical for at-scale execution with our middleware, so the provenance described above will need to be saved to files at first and loaded into the database later.
+
+Most of the information we need to save is already included in the ``QuantumGraph`` produced prior to execution, especially if we include UUIDs for its predicted intermediate and output in the graph at construction.
```

Comment:
If UUIDs are saved in the graph at construction, we'll need other mechanisms for later creating new UUIDs if allowing folks to use same QuantumGraph (or Quantum Provenance) to execute a new run.